client/allocrunnerv2/alloc_runner.go

```diff
@@ -385,6 +388,27 @@ func getClientStatus(taskStates map[string]*structs.TaskState) (status, descript
 	return "", ""
 }
 
+// AllocState returns a copy of allocation state including a snapshot of task
+// states.
+func (ar *allocRunner) AllocState() *state.State {
+	// Must acquire write-lock in case TaskStates needs to be set.
+	ar.stateLock.Lock()
+	defer ar.stateLock.Unlock()
+
+	// If TaskStateUpdated has not been called yet, ar.state.TaskStates
+	// won't be set as it is not the canonical source of TaskStates.
+	if len(ar.state.TaskStates) == 0 {
+		ar.tasksLock.RLock()
+		ar.state.TaskStates = make(map[string]*structs.TaskState, len(ar.tasks))
```

Comment:
This is mutating the state. Should you call the copy first and then do the assignment on the copy.

Reply:
Done!